src/include/daos_fs_sys.h

```diff
@@ -0,0 +1,506 @@
+ * \param[in]	flags	Access flags (handles: O_RDONLY, O_RDWR, O_EXCL,
+ *			O_CREAT, O_TRUNC, O_NOFOLLOW).
+ * \param[in]	cid	DAOS object class id (pass 0 for default MAX_RW).
+ *			Valid on create only; ignored otherwise.
+ * \param[in]	chunk_size
+ *			Chunk size of the array object to be created.
+ *			(pass 0 for default 1 MiB chunk size).
+ *			Valid on file create only; ignored otherwise.
+ * \param[in]	value	Symlink value (NULL if not syml).
+ * \param[out]	obj	Pointer to object opened.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_open(dfs_sys_t *dfs_sys, const char *path, mode_t mode, int flags,
+	     daos_oclass_id_t cid, daos_size_t chunk_size,
+	     const char *value, dfs_obj_t **obj);
+
+/**
+ * Close/release open object.
+ *
+ * \param[in]	obj	Object to release.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_close(dfs_obj_t *obj);
+
+/**
+ * Read data from the file object, and return actual data read.
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	obj	Opened file object.
+ * \param[in,out]
+ *		buf	[in]: Allocated buffer for data.
+ *			[out]: Actual data read.
+ * \param[in]	off	Offset into the file to read from.
+ * \param[in,out]
+ *		size	[in]: Size of buffer passed in. [out]: Actual size of
+ *			      data read. On error, this is set to -1.
+ * \param[in]	ev	Completion event, it is optional and can be NULL.
+ *			Function will run in blocking mode if \a ev is NULL.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_read(dfs_sys_t *dfs_sys, dfs_obj_t *obj, void *buf, daos_off_t off,
+	     daos_size_t *size, daos_event_t *ev);
+
+/**
+ * Write data to the file object.
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	obj	Opened file object.
+ * \param[in]	buf	Data to write.
+ * \param[in]	off	Offset into the file to write to.
+ * \param[in,out]
+ *		size	[in]: Size of buffer passed in. [out]: Actual size of
+ *			      data written. On error, this is set to -1.
+ * \param[in]	ev	Completion event, it is optional and can be NULL.
+ *			Function will run in blocking mode if \a ev is NULL.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_write(dfs_sys_t *dfs_sys, dfs_obj_t *obj, const void *buf,
+	      daos_off_t off, daos_size_t *size, daos_event_t *ev);
+
+/**
+ * Punch a hole in the file starting at offset to len. If len is set to
+ * DFS_MAX_FSIZE, this will be a truncate operation to punch all bytes in the
+ * file above offset. If the file size is smaller than offset, the file is
+ * extended to offset and len is ignored.
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	path	Link path of file.
+ * \param[in]	offset	offset of file to punch at.
+ * \param[in]	len	number of bytes to punch.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_punch(dfs_sys_t *dfs_sys, const char *path,
```

Comment:
do we need punch on a path and can it just be on an open obj?
like read / write..

Reply:
It's not strictly necessary, since a user could always just open and punch themselves. It's more of a shortcut. mpifileutils does use this, but it could instead call open + obj punch.
Do you think we **shouldn't** have this function?

Reply:
no that's fine i guess if mpifileutils is using this.
do we need then the other one opunch?

Reply:
Oops, no. I meant to remove this after adding `dfs_sys2base`

Reply:
i approved this, but let's remove the opunch in this case

Reply:
Sure, done!